src/grpc/infer_handler.h

```diff
@@ -652,12 +651,143 @@ class InferHandlerState {
+              state->irequest_ptr_, &is_cancelled);
+          if ((err == nullptr) && (!is_cancelled)) {
+            err = TRITONSERVER_InferenceRequestCancel(state->irequest_ptr_);
+          }
+          if (err != nullptr) {
+            const char* request_id = "";
+            TRITONSERVER_Error* tmp_err = TRITONSERVER_InferenceRequestId(
+                state->irequest_ptr_, &request_id);
+            if (tmp_err != nullptr) {
+              // Ignore if failed to get request id.
+              TRITONSERVER_ErrorDelete(tmp_err);
+            }
+            TRITONSERVER_Error* wrapped_err = TRITONSERVER_ErrorNew(
+                TRITONSERVER_ERROR_INTERNAL,
+                std::string(
+                    std::string("[request id:") + request_id +
+                    "]: " + TRITONSERVER_ErrorMessage(err))
+                    .c_str());
+            TRITONSERVER_ErrorDelete(err);
+            return wrapped_err;
+          }
+
+          {
+            std::lock_guard<std::recursive_mutex> lock(state->step_mtx_);
+            if (state->step_ != Steps::CANCELLED) {
+              state->step_ = Steps::CANCELLATION_ISSUED;
+            }
+          }
+        }
+      }
+      return nullptr;
+    }
+
+
+    // Handles the gRPC context cancellation. This function can be called
+    // multiple times and is supposed to be re-entrant.
+    // Returns whether or not to continue cycling through the gRPC
+    // completion queue or not.
+    bool HandleCancellation(
+        InferHandlerStateType* state, bool rpc_ok, const std::string& name)
+    {
+      if (!IsCancelled()) {
+        LOG_ERROR
+            << "[INTERNAL] HandleCancellation called even when the context was "
+               "not cancelled for "
+            << name << ", rpc_ok=" << rpc_ok << ", context "
+            << state->context_->unique_id_ << ", " << state->unique_id_
+            << " step " << state->step_;
+        return true;
+      }
+      if ((state->step_ != Steps::CANCELLATION_ISSUED) &&
+          (state->step_ != Steps::CANCELLED)) {
+        LOG_VERBOSE(1) << "Cancellation notification received for " << name
+                       << ", rpc_ok=" << rpc_ok << ", context "
+                       << state->context_->unique_id_ << ", "
+                       << state->unique_id_ << " step " << state->step_;
+
+        // If the context has not been cancelled then
+        // issue cancellation request to all the inflight
+        // states belonging to the context.
+        if (state->context_->step_ != Steps::CANCELLED) {
+          TRITONSERVER_Error* err = nullptr;
+          err = IssueRequestCancellation();
+          if (err != nullptr) {
+            LOG_VERBOSE(1) << "Failed to cancel request: "
+                           << TRITONSERVER_ErrorMessage(err);
+            TRITONSERVER_ErrorDelete(err);
+          }
+          // Mark the context as cancelled
+          state->context_->step_ = Steps::CANCELLED;
```

Comment:
Is it possible for the `step` to be modified by another thread? Do we need to project it using a mutex?

Reply:
We only have a single ModelStreamInferHandler thread. Hence, we don't need a lock around the context step_.